In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
#no of variable
N = 12

In [ ]:
#no of constraint
M = 5

In [ ]:
#create model
model2 = ConcreteModel()

In [ ]:
# declare objective coefficient
obj_coef_c = np.array([0.56,1.721,-9.6,-2.85,41.5,33.05,-0.75,20.5,-0.35,-31.02,50.65,6.125])

In [ ]:
#declare constraint coefficient
constr_coef_A = np.array([[1,-1,0,0,2,-2,1,0,0,-1,1,-2],[1,0,0,-1,0,-2,0,-1,1,0,0,-1],[-1,0,0,1,-1,2.5,-3.5,0,2,-1,1,0],[0,1,-0.45,0,-2.78,0,-3.4,0,-1,1,-1,0.25],[0,-1,1,0,1,0,1,3,-2,0,1,0]])

In [ ]:
constr_rhs_b = [88.5,35.4,47.9,41.43,22.7]

In [ ]:
col_indices = np.arange(N)

In [ ]:
row_indices = np.arange(M)

In [ ]:
#lower and upper bound
lower_bound = np.array([0,0,-np.inf,-np.inf,-34,0,-67,0,-23,-5,6,1])
upper_bound = np.array([12,503,4,np.inf,np.inf,2,20,55,23,15,np.inf,np.inf])

In [ ]:
model2.x=Var(col_indices)

In [ ]:
model2.constraints=ConstraintList()

In [ ]:
for j in row_indices:
  model2.constraints.add(sum(model2.x[i]*constr_coef_A[j][i] for i in col_indices) <= constr_rhs_b[j])

In [ ]:
model2.objective = Objective( expr = sum(model2.x[i]*obj_coef_c[i] for i in col_indices),sense=minimize)

In [ ]:
#set lower bound and upper bound for variable
for i in col_indices:
  model2.x[i].setlb(lower_bound[i])
  model2.x[i].setub(upper_bound[i])

In [ ]:
model2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

1 Var Declarations
    x : Size=12, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :   0.0 :  None :  12.0 : False :  True :  Reals
          1 :   0.0 :  None : 503.0 : False :  True :  Reals
          2 :  None :  None :   4.0 : False :  True :  Reals
          3 :  None :  None :  None : False :  True :  Reals
          4 : -34.0 :  None :  None : False :  True :  Reals
          5 :   0.0 :  None :   2.0 : False :  True :  Reals
          6 : -67.0 :  None :  20.0 : False :  True :  Reals
          7 :   0.0 :  None :  55.0 : False :  True :  Reals
          8 : -23.0 :  None :  23.0 : False 

In [ ]:
!apt-get install -q -yy glpk-utils

Reading package lists...
Building dependency tree...
Reading state information...
glpk-utils is already the newest version (4.65-1).
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 19 not upgraded.


In [ ]:
SolverFactory('glpk',excutable='/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1951.831
  Upper bound: -1951.831
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 13
  Number of nonzeros: 38
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01495218276977539
# ----------------------------------------------------------
#   Solution Information
# ---------------------------

In [ ]:
from pyomo.core.base import constraint
print('\n objective value',model2.objective())
print('\n variables')
for i in col_indices:
  print('x[',i,'] = ',model2.x[i].value)
model2.constraints.display()



 objective value -1951.8309999999997

 variables
x[ 0 ] =  12.0
x[ 1 ] =  0.0
x[ 2 ] =  4.0
x[ 3 ] =  119.82
x[ 4 ] =  -34.0
x[ 5 ] =  0.0
x[ 6 ] =  20.0
x[ 7 ] =  0.0
x[ 8 ] =  -7.46000000000002
x[ 9 ] =  15.0
x[ 10 ] =  6.0
x[ 11 ] =  1.0
constraints : Size=5
    Key : Lower : Body                : Upper
      1 :  None :               -47.0 :  88.5
      2 :  None : -116.28000000000002 :  35.4
      3 :  None :   47.89999999999995 :  47.9
      4 :  None :   41.43000000000002 : 41.43
      5 :  None :   10.92000000000004 :  22.7


$ \Huge{solution \ \ of \ \ 2.1} \\ \\ \huge{optimal \ \ value \ \ of \ \ objective \ \ function = -1951.8309999999997} \\ \large{optimal \ \ value \ \ of  \ \ decision \ \ \ variable } \\  x_1 =12.0 \\ x_2 = 0.00 \\ x_3 = 4.0 \\ x_4 = 119.82 \\ x_5 = -34.0  \\ x_6 = 0.00 \\ x_7 = 20.0  \\ x_8 =0.00 \\ x_9 = -7.46000000000002 \\ x_10 = 15.0 \\ x_11 = 6.0 \\ x_12 = 1.0$

$ \Huge{solution \ \ of \ \ 2.2} \\ \\ \text{constraint1  is  inactive  since  constraint1  < 88.5} \\ \text{constraint2  is  inactive  since  constraint2 < 35.4} \\ \text{constraint3  is  active  since  constraint3 =47.9} \\ \text{constraint4  is  active  since  constraint4 = 41.43} \\ \text{constraint5  is  inactive  since  constraint5 < 22.7} $